File: libraries/billing/src/commonMain/kotlin/com/sodogku/libraries/billing/StoreBilling.kt
@file:OptIn(ExperimentalObjCName::class)

package com.sodogku.libraries.billing

import kotlin.experimental.ExperimentalObjCName
import kotlin.native.ObjCName

/**
 * The store's answer to "does this device own the product".
 *
 * [Unknown] is the whole reason this is three-valued. A network failure, a
 * signed-out Play account or a StoreKit hiccup all mean *we could not ask*, and
 * a two-valued answer would collapse that into "not owned" — which is how a
 * paying customer starts seeing ads on a train. Only [NotOwned] is allowed to
 * clear the cached entitlement.
 */
@ObjCName("StoreOwnership", exact = true)
enum class StoreOwnership {
    Owned,
    NotOwned,
    Unknown,
}

/** How a purchase attempt ended, flattened for the Swift side. */
@ObjCName("StorePurchaseResult", exact = true)
enum class StorePurchaseResult {
    Purchased,
    Cancelled,
    AlreadyOwned,

    /** Billing is unavailable on this device: no Play Services, parental controls, sandbox off. */
    Unavailable,
    Failed,
}

/** A result plus a machine-readable failure kind for `iap.purchase_result`. */
@ObjCName("StorePurchaseOutcome", exact = true)
class StorePurchaseOutcome(
    val result: StorePurchaseResult,
    val errorKind: String? = null,
)

/**
 * The thin platform seam under [Entitlements]: talk to Play or StoreKit, say
 * what the store said. No caching, no policy, no state.
 *
 * **Android binding** is `PlayStoreBilling` in `:libraries:billing:impl/androidMain`
 * (Play Billing). **iOS binding** is `IOSStoreBilling` (Swift, StoreKit 2) handed
 * to the graph through `IosAppComponent`.
 *
 * Implementations must not throw — a store that is down has to arrive here as
 * [StoreOwnership.Unknown] or [StorePurchaseResult.Failed], because the layer
 * above treats a thrown exception and a "no" identically and only one of those
 * is safe.
 */
@ObjCName("StoreBilling", exact = true)
interface StoreBilling {

    /**
     * Ask the store whether [productId] is owned right now.
     *
     * On Android this is `queryPurchasesAsync`, which reads Play's local cache
     * and refreshes it — it works offline for a device that has seen the
     * purchase before. On iOS it is `Transaction.currentEntitlements`, which is
     * likewise locally verified.
     */
    suspend fun ownership(productId: String): StoreOwnership

    /** Runs the purchase flow. Must be safe to call when the product is already owned. */
    suspend fun purchase(productId: String): StorePurchaseOutcome

    /**
     * The explicit restore Apple requires a visible control for. On Android
     * there is nothing to restore — `queryPurchasesAsync` already is the
     * restore — so the Play implementation forwards to [ownership].
     */
    suspend fun restore(productId: String): StoreOwnership

    /**
     * Localised price as the store formats it ("$4.99", "4,99 €"), or null when
     * the store could not be reached. Never hardcode a price: it is set per
     * storefront and SPEC 4.3 keeps it out of config for the same reason.
     */
    suspend fun priceLabel(productId: String): String?
}

/**
 * The store product. One non-consumable, SPEC 5.1.
 *
 * Same string on both stores by design — Play calls it a managed product, Apple
 * a non-consumable, and keeping the ids identical means one constant rather than
 * a platform branch on every call site. It is **not** in remote config: SPEC 4.4
 * lists the product id as a store operation.
 */
@ObjCName("ProductIds", exact = true)
object ProductIds {
    const val pro: String = "sodogku_pro"
}

File: apps/ios/iosApp/Platform/StoreBilling.swift
//
//  StoreBilling.swift
//  iosApp
//
//  StoreKit 2. No SDK to add — StoreKit ships with the OS — so unlike
//  `AdNetwork.swift` this is unguarded and real from the first build.
//
//  The one rule that matters is the same one the Kotlin side is built around:
//  a store we could not reach reports `.unknown`, never `.notOwned`. Only an
//  answer clears a paying customer's entitlement.
//
//  ── SETUP REQUIRED ───────────────────────────────────────────────────────
//  Local testing needs a StoreKit configuration file, because a product id
//  that App Store Connect has never heard of returns an empty product list and
//  the paywall shows no price:
//
//    File ▸ New ▸ File ▸ StoreKit Configuration File (Products.storekit)
//    Add a Non-Consumable with product id `sodogku_pro` (ProductIds.pro),
//    price 4.99, then Product ▸ Scheme ▸ Edit Scheme ▸ Run ▸ Options ▸
//    StoreKit Configuration ▸ Products.storekit.
//
//  For a sandbox test on a device, create the same non-consumable in App Store
//  Connect and sign in with a Sandbox Apple Account.
//
import ComposeApp
import Foundation
import StoreKit

class IOSStoreBilling: StoreBilling {

    func __ownership(productId: String) async throws -> StoreOwnership {
        await currentOwnership(of: productId)
    }

    /// Apple requires a visible restore control, and this is what it calls.
    /// `AppStore.sync()` forces a receipt refresh (it can prompt for the Apple
    /// Account password), which is the difference between "restore" and the
    /// ordinary entitlement check above.
    func __restore(productId: String) async throws -> StoreOwnership {
        do {
            try await AppStore.sync()
        } catch {
            // A cancelled password prompt or a network failure. We still ask
            // for the local entitlements below — they may already be there.
            NSLog("StoreKit sync failed during restore: \(error)")
        }
        return await currentOwnership(of: productId)
    }

    func __purchase(productId: String) async throws -> StorePurchaseOutcome {
        guard let product = await product(for: productId) else {
            return StorePurchaseOutcome(result: .unavailable, errorKind: "product_not_found")
        }

        do {
            switch try await product.purchase() {
            case .success(let verification):
                switch verification {
                case .verified(let transaction):
                    await transaction.finish()
                    return StorePurchaseOutcome(result: .purchased, errorKind: nil)
                case .unverified(_, let error):
                    // A failed signature check. Treat it as a failure rather
                    // than a grant: this is the one place where being generous
                    // would be giving the product away to a jailbroken device.
                    return StorePurchaseOutcome(result: .failed, errorKind: "unverified_\((error as NSError).code)")
                }
            case .userCancelled:
                return StorePurchaseOutcome(result: .cancelled, errorKind: nil)
            case .pending:
                // Ask to Buy, or a payment awaiting approval. Not owned yet;
                // the foreground refresh in `RealEntitlements` picks it up.
                return StorePurchaseOutcome(result: .failed, errorKind: "purchase_pending")
            @unknown default:
                return StorePurchaseOutcome(result: .failed, errorKind: "unknown_purchase_result")
            }
        } catch StoreKitError.userCancelled {
            return StorePurchaseOutcome(result: .cancelled, errorKind: nil)
        } catch {
            return StorePurchaseOutcome(result: .failed, errorKind: "storekit_\((error as NSError).code)")
        }
    }

    func __priceLabel(productId: String) async throws -> String? {
        await product(for: productId)?.displayPrice
    }

    // MARK: - Private

    private func currentOwnership(of productId: String) async -> StoreOwnership {
        // `currentEntitlements` is served from the on-device receipt, so a
        // device that has seen the purchase answers with no network at all.
        // That is what keeps an offline Pro player Pro.
        for await result in Transaction.currentEntitlements {
            guard case .verified(let transaction) = result else { continue }
            if transaction.productID == productId && transaction.revocationDate == nil {
                return .owned
            }
        }

        // Nothing locally — and StoreKit gives no "could not reach the store"
        // signal here, so an empty result on a device signed out of the App
        // Store looks exactly like a genuine "never bought it". A catalogue
        // lookup is the cheapest available proof that we could talk to the
        // store at all; if that fails too then we do not know, and not knowing
        // must never clear a paying customer.
        let storeAnswered = await product(for: productId) != nil
        return storeAnswered ? .notOwned : .unknown
    }

    private func product(for productId: String) async -> Product? {
        do {
            return try await Product.products(for: [productId]).first
        } catch {
            NSLog("StoreKit product lookup failed: \(error)")
            return nil
        }
    }
}
